// @marko/compiler/dist/index.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");exports.__esModule = true;exports.configure = configure;exports.compile = compile;exports.compileSync = compileSync;exports.compileFile = compileFile;exports.compileFileSync = compileFileSync;exports.getRuntimeEntryFiles = getRuntimeEntryFiles;exports._clearDefaults = _clearDefaults;exports.taglib = exports.types = void 0;var _types = _interopRequireWildcard(require("./babel-types"));exports.types = _types;
var babel = _interopRequireWildcard(require("@babel/core"));
var _babelPlugin = _interopRequireDefault(require("./babel-plugin"));
var _config = _interopRequireDefault(require("./config"));
var taglib = _interopRequireWildcard(require("./taglib"));exports.taglib = taglib;
var _shouldOptimize = _interopRequireDefault(require("./util/should-optimize"));
var _tryLoadTranslator = _interopRequireDefault(require("./util/try-load-translator"));


let globalConfig = { ..._config.default };
function configure(newConfig) {
  globalConfig = { ..._config.default, ...newConfig };
}

async function compile(src, filename, config) {
  const babelConfig = loadBabelConfig(filename, config);
  const babelResult = await babel.transformAsync(src, babelConfig);
  scheduleDefaultClear(config);
  return buildResult(babelResult);
}

function compileSync(src, filename, config) {
  const babelConfig = loadBabelConfig(filename, config);
  const babelResult = babel.transformSync(src, babelConfig);
  scheduleDefaultClear(config);
  return buildResult(babelResult);
}

async function compileFile(filename, config) {
  return new Promise((resolve, reject) => {
    getFs(config).readFile(filename, "utf-8", (err, src) => {
      if (err) {
        return reject(err);
      }

      return resolve(compile(src, filename, config));
    });
  });
}

function compileFileSync(filename, config) {
  const src = getFs(config).readFileSync(filename, "utf-8");
  return compileSync(src, filename, config);
}

function getRuntimeEntryFiles(output, requestedTranslator) {
  const translator = (0, _tryLoadTranslator.default)(requestedTranslator);
  if (translator && translator.getRuntimeEntryFiles) {
    return translator.getRuntimeEntryFiles(output, (0, _shouldOptimize.default)());
  }

  return [];
}

function loadBabelConfig(filename, config) {
  const markoConfig = { ...globalConfig, ...config, babelConfig: undefined };
  const requiredPlugins = [[_babelPlugin.default, markoConfig]];
  const baseBabelConfig = {
    ...(config && config.babelConfig),
    filename,
    sourceFileName: filename,
    sourceType: "module",
    sourceMaps: markoConfig.sourceMaps,
    code: markoConfig.code,
    ast: markoConfig.ast };


  if (markoConfig.modules === "cjs") {
    requiredPlugins.push([
    require.resolve("@babel/plugin-transform-modules-commonjs"),
    { loose: true }]);

  }

  baseBabelConfig.plugins = requiredPlugins.concat(
  baseBabelConfig.plugins || []);


  return babel.loadPartialConfig(baseBabelConfig).options;
}

function buildResult(babelResult) {
  const {
    ast,
    map,
    code,
    metadata: { marko: meta } } =
  babelResult;
  return { ast, map, code, meta };
}

let scheduledClear = false;
let clearingDefaultFs = false;
let clearingDefaultCache = false;
function scheduleDefaultClear(config) {
  if (!scheduledClear) {
    clearingDefaultCache = isDefaultCache(config);
    clearingDefaultFs = isDefaultFS(config);

    if (clearingDefaultCache || clearingDefaultFs) {
      scheduledClear = true;
      setImmediate(_clearDefaults);
    }
  }
}

function _clearDefaults() {
  if (clearingDefaultCache) {
    clearingDefaultCache = false;
    globalConfig.cache.clear();
  }

  if (clearingDefaultFs) {
    clearingDefaultFs = false;
    globalConfig.fileSystem.purge();
  }
  scheduledClear = false;
}

function isDefaultCache(config) {
  return !config.cache || config.cache === globalConfig.cache;
}

function isDefaultFS(config) {
  return getFs(config) === globalConfig.fileSystem;
}

function getFs(config) {
  return config.fileSystem || globalConfig.fileSystem;
}
